src/functions.ts

```diff
@@ -137,3 +148,79 @@ export interface CloudFunctionsContext {
  * @public
  */
 export type Context = CloudFunctionsContext | CloudEvent<unknown>;
+
+/**
+ * InvocationRequest represents the properties of an invocation over HTTP.
+ * @public
+ */
+export interface InvocationRequest {
+  /** Returns the request body as either a string or a Buffer if the body is binary. */
+  body(): string | Buffer;
+  /** Header returns the value of the specified header */
+  header(header: string): string | undefined;
+}
+
+/**
+ * InvocationResponse interface describes the properties that can be set on
+ * an invocation response.
+ * @public
+ */
+export interface InvocationResponse {
+  /** Sets a header on the response. */
+  setHeader(key: string, value: string): void;
+  /** Writes a chunk of data to the response. */
+  write(data: string | Buffer): void;
+  /** Ends the response, must be called once at the end of writing. */
+  end(data: string | Buffer): void;
+}
+
+/**
+ * The contract for a request deserializer and response serializer.
+ * @public
+ */
+export interface InvocationFormat<T, U> {
+  /**
+   * Creates an instance of the request type from an invocation request.
+   *
+   * @param request the request body as raw bytes
+   */
+  deserializeRequest(request: InvocationRequest): T | Promise<T>;
+
+  /**
+   * Writes the response type to the invocation result.
+   *
+   * @param responseWriter interface for writing to the invocation result
+   * @param response the response object
+   */
+  serializeResponse(
+    responseWriter: InvocationResponse,
+    response: U
+  ): void | Promise<void>;
+}
+
+/**
+ * Default invocation format for JSON requests.
+ * @public
+ */
+export class JsonInvocationFormat<T, U> implements InvocationFormat<T, U> {
```

Comment:
See below, my thinking is to disable body-parser JSON middleware for typed requests to allow the InvocationFormat to intercept and handle the request since it should be user customizable.

An alternative would be to leave the body parser in place and simply change the `InvocationRequest.body()` signature to expose `string | Buffer | object`?